src/System/Random/Internal.hs

```diff
@@ -0,0 +1,1011 @@
+{-# INLINE pinnedByteArrayToForeignPtr #-}
+
+
+-- | Generates a pseudo-random 'ByteString' of the specified size.
+--
+-- @since 1.2
+uniformByteString :: StatefulGen g m => Int -> g -> m ByteString
+uniformByteString n g = do
+  ba@(SBS ba#) <- uniformShortByteString n g
+  pure $
+    if isTrue# (isByteArrayPinned# ba#)
+      then pinnedByteArrayToByteString ba#
+      else fromShort ba
+{-# INLINE uniformByteString #-}
+
+
+-- | Opaque data type that carries the type of a pure pseudo-random number
+-- generator.
+--
+-- @since 1.2
+data StateGenM g = StateGenM
+
+-- | Wrapper for pure state gen, which acts as an immutable seed for the corresponding
+-- stateful generator `StateGenM`
+--
+-- @since 1.2
+newtype StateGen g = StateGen g
+
+instance (RandomGen g, MonadState g m) => StatefulGen (StateGenM g) m where
+  uniformWord32R r _ = state (genWord32R r)
+  uniformWord64R r _ = state (genWord64R r)
+  uniformWord8 _ = state genWord8
+  uniformWord16 _ = state genWord16
+  uniformWord32 _ = state genWord32
+  uniformWord64 _ = state genWord64
+  uniformShortByteString n _ = state (genShortByteString n)
+
+instance (RandomGen g, MonadState g m) => FrozenGen (StateGen g) m where
+  type MutableGen (StateGen g) m = StateGenM g
+  freezeGen _ = fmap StateGen get
+  thawGen (StateGen g) = StateGenM <$ put g
+
+-- | Splits a pseudo-random number generator into two. Updates the state with
+-- one of the resulting generators and returns the other.
+--
+-- @since 1.2
+splitGen :: (MonadState g m, RandomGen g) => m g
+splitGen = state split
+
+-- | Runs a monadic generating action in the `State` monad using a pure
+-- pseudo-random number generator.
+--
+-- @since 1.2
+runStateGen :: RandomGen g => g -> (StateGenM g -> State g a) -> (a, g)
+runStateGen g f = runState (f StateGenM) g
+
+-- | Runs a monadic generating action in the `State` monad using a pure
+-- pseudo-random number generator. Returns only the resulting pseudo-random
+-- value.
+--
+-- @since 1.2
+runStateGen_ :: RandomGen g => g -> (StateGenM g -> State g a) -> a
+runStateGen_ g = fst . runStateGen g
+
+-- | Runs a monadic generating action in the `StateT` monad using a pure
+-- pseudo-random number generator.
+--
+-- @since 1.2
+runStateGenT :: RandomGen g => g -> (StateGenM g -> StateT g m a) -> m (a, g)
+runStateGenT g f = runStateT (f StateGenM) g
+
+-- | Runs a monadic generating action in the `StateT` monad using a pure
+-- pseudo-random number generator. Returns only the resulting pseudo-random
+-- value.
+--
+-- @since 1.2
+runStateGenT_ :: (RandomGen g, Functor f) => g -> (StateGenM g -> StateT g f a) -> f a
+runStateGenT_ g = fmap fst . runStateGenT g
+
+-- | Runs a monadic generating action in the `ST` monad using a pure
+-- pseudo-random number generator.
+--
+-- @since 1.2
+runStateGenST :: RandomGen g => g -> (forall s . StateGenM g -> StateT g (ST s) a) -> (a, g)
+runStateGenST g action = runST $ runStateGenT g action
```

Comment:
It definitely would ;) But we do not yet have impredicative polymorphism. I just watched a talk from SPJ about a week ago and learned that the only reason current code works is due to special casing  in ghc for `$`. In case you are curious: https://www.youtube.com/watch?v=ZuNMo136QqI

Reply:
As illustration:
```
λ> runST $ return "X"
"X"
λ> runST `id` return "X"
<interactive>:3:1: error:
    • Couldn't match type ‘forall s. ST s t’ with ‘m0 [Char]’
      Expected type: m0 [Char] -> t
        Actual type: (forall s. ST s t) -> t
...
```